app/services/search/postgres/reading_list.rb

```diff
@@ -58,12 +59,20 @@ def self.search_documents(user, term: nil, statuses: [], tags: [], page: 0, per_
         }
       end
 
-      def self.find_articles(user_id:, term:, statuses:, tags:, page:, per_page:)
-        relation = ::Article
-          .joins(:reactions)
-          .where("reactions.category": :readinglist)
-          .where("reactions.user_id": user_id)
-          .where("reactions.status": statuses)
+      def self.find_articles(user:, term:, statuses:, tags:, page:, per_page:)
+        # [@jgaskins, @rhymes] as `reactions` is potentially a big table, adding pagination
+        # to an INNER JOIN (eg. `joins(:reactions)`) exponentially decreases the performance,
+        # incrementing query time as the database has to scan all the rows just to discard
+        # them right after if they lie outside the bounds of the `OFFSET`.
+        # Even though it should have had a similar performance, we realized that a subquery
+        # enabled PostgreSQL query planner to drastically decrease the planned time (ca. 145x)
+        reaction_query_sql = user.reactions.readinglist
+          .where(status: statuses, reactable_type: "Article")
+          .order(created_at: :desc)
+          .select(*REACTION_ATTRIBUTES)
+          .to_sql
+
+        relation = Article.joins("INNER JOIN (#{reaction_query_sql}) reactions ON reactions.reactable_id = articles.id")
```

Comment:
This is a pretty baller move! When we talked about this before, I was wondering how you were going to do that subquery.

I don't think this should be a super common idiom (I had some trouble at my previous company where a few people got pretty clever with combining queries like this all over the app and it caused issues with prepared statements and replication), but if we keep these sparse it can be a really nice solution for some of the query-perf issues!

Reply:
Ah ah thank DuckDuckGo, it took me a bit of reading on how to create this SQL query. I wish the Rails team would re-publish `ARel` as a public interface like when it was a separate gem, it was so much easier to write complex queries with it 😢